ding/policy/ibc.py

```diff
@@ -0,0 +1,165 @@
+from typing import Dict, Any, List, Tuple
+from collections import namedtuple
+from easydict import EasyDict
+
+import torch
+import torch.nn.functional as F
+
+from easydict import EasyDict
+
+from ding.model import model_wrap
+from ding.torch_utils import to_device
+from ding.utils.data import default_collate, default_decollate
+from ding.utils import POLICY_REGISTRY
+from .bc import BehaviourCloningPolicy
+from ding.model.template.ebm import create_stochastic_optimizer
+from ding.model.template.ebm import StochasticOptimizer, MCMC, AutoRegressiveDFO
+from ding.torch_utils import unsqueeze_repeat
+
+
+@POLICY_REGISTRY.register('ibc')
+class IBCPolicy(BehaviourCloningPolicy):
+
+    config = dict(
+        type='ibc',
+        cuda=False,
+        on_policy=False,
+        continuous=True,
+        learn=dict(
+            train_epoch=30000,
+            batch_size=256,
+            multi_gpu=False,
+            optim=dict(
+                learning_rate=1e-5,
+                weight_decay=0.0,
+                beta1=0.9,
+                beta2=0.999,
+            ),
+        ),
+        collect=dict(normalize_states=True,),
+        eval=dict(evaluator=dict(eval_freq=10000, )),
+    )
+
+    def default_model(self) -> Tuple[str, List[str]]:
+        return 'ebm', ['ding.model.template.ebm']
+    
+    def _init_learn(self):
+        optim_cfg = self._cfg.learn.optim
+        self._optimizer = torch.optim.Adam(
+            self._model.parameters(),
+            lr=optim_cfg.learning_rate,
+            weight_decay=optim_cfg.weight_decay,
+            betas=(optim_cfg.beta1, optim_cfg.beta2),
+        )
+        self._stochastic_optimizer: StochasticOptimizer = \
+            create_stochastic_optimizer(self._cfg.model.stochastic_optim)
+        self._learn_model = model_wrap(self._model, 'base')
+        self._learn_model.reset()
+    
+    def _forward_learn(self, data):
+        if not isinstance(data, dict):
+            data = default_collate(data)
+        if self._cuda:
+            data = to_device(data, self._device)
+        self._learn_model.train()
+
+        loss_dict = dict()
+
+        # obs: (B, O)
+        # action: (B, A)
+        obs, action = data['obs'], data['action']
+
+        # (B, N, O), (B, N, A)
+        obs, negatives = self._stochastic_optimizer.sample(obs, self._learn_model)
+
+        # (B, N+1, A)
+        targets = torch.cat([action.unsqueeze(dim=1), negatives], dim=1)
+        # (B, N+1, O)
+        obs = torch.cat([obs[:, :1], obs], dim=1)
```

Comment:
can we merge the infoNCE loss into [this implementation](https://github.com/opendilab/DI-engine/blob/main/ding/torch_utils/loss/contrastive_loss.py)

Reply:
I am not sure. But I think we can create another PR in the future to change both algorithms.

Reply:
ok, add this item to to-do list